internal/task/ondemandsnapshot/ondemandsnapshot.go

```diff
@@ -0,0 +1,253 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// ErrGetEtcd represents the error in case of fetching etcd object. Reason could be internal error
+	ErrGetEtcd v1alpha1.ErrorCode = "ERR_GET_ETCD"
+	//  ErrEtcdNotReady represents the error in case etcd is not ready
+	ErrEtcdNotReady v1alpha1.ErrorCode = "ERR_ETCD_NOT_READY"
+	//  ErrBackupNotEnabled represents the error in case backup is not enabled
+	ErrBackupNotEnabled v1alpha1.ErrorCode = "ERR_BACKUP_NOT_ENABLED"
+	// ErrGetSecret represents the error in case of failure in fetching the secret
+	ErrGetSecret v1alpha1.ErrorCode = "ERR_GET_SECRET"
+	// ErrCreateHTTPRequest represents the error in case of failure in creating http request
+	ErrCreateHTTPRequest v1alpha1.ErrorCode = "ERR_CREATE_HTTP_REQUEST"
+	// ErrExecuteHTTPRequest represents the error in case of failure in executing http request
+	ErrExecuteHTTPRequest v1alpha1.ErrorCode = "ERR_EXECUTE_HTTP_REQUEST"
+	//  ErrCreateSnapshot represents the error in case of failure in creating snapshot
+	ErrCreateSnapshot v1alpha1.ErrorCode = "ERR_CREATE_SNAPSHOT"
+)
+
+// OnDemandSnapshotTask implements the task.Handler interface for handling on-demand snapshot tasks.
+type OnDemandSnapshotTask struct {
+	client        client.Client
+	logger        logr.Logger
+	name          string
+	etcdReference v1alpha1.EtcdReference
+	httpClient    http.Client
+	config        v1alpha1.OnDemandSnapshotConfig
+}
+
+// New creates a new instance of OnDemandSnapshotTask.
+func New(k8sclient client.Client, logger logr.Logger, task *v1alpha1.EtcdOpsTask) (task.Handler, error) {
+	return NewWithHTTPClient(k8sclient, logger, task, nil)
+}
+
+// NewWithHTTPClient creates a new instance of OnDemandSnapshotTask with a custom HTTP client.
+func NewWithHTTPClient(k8sclient client.Client, logger logr.Logger, task *v1alpha1.EtcdOpsTask, httpClient *http.Client) (task.Handler, error) {
+	etcdRef := *task.Spec.EtcdRef
+	if etcdRef.Namespace == "" {
+		etcdRef.Namespace = task.Namespace
+	}
+
+	// Use provided HTTP client or create default
+	var client http.Client
+	if httpClient != nil {
+		client = *httpClient
+	} else {
+		client = http.Client{Timeout: time.Second * time.Duration(*task.Spec.Config.OnDemandSnapshot.TimeoutSeconds)}
+	}
+
+	return &OnDemandSnapshotTask{
+		client:        k8sclient,
+		logger:        logger,
+		name:          task.Name,
+		etcdReference: etcdRef,
+		httpClient:    client,
+		config:        *task.Spec.Config.OnDemandSnapshot,
+	}, nil
+}
+
+// EtcdReference returns the NamespacedName of the etcd object referenced by the task.
+func (o *OnDemandSnapshotTask) EtcdReference() types.NamespacedName {
+	return types.NamespacedName{
+		Name:      o.etcdReference.Name,
+		Namespace: o.etcdReference.Namespace,
+	}
+}
+
+// Name returns the name of the task.
+func (o *OnDemandSnapshotTask) Name() string {
+	return o.name
+}
+
+// Logger returns the logger for the task.
+func (o *OnDemandSnapshotTask) Logger() logr.Logger {
```

Comment:
as commented above by different reviewers, this is not required, please remove it.